### PR TITLE
Unify CT-INT vertex selection with RB tree

### PR DESCRIPTION
**Pull request by gbalduzz**
~~Depends on #191.~~
Implemented a (rather fancy) structure to insert, remove and sample CT-INT vertices, removing the need for a fast or reproducible code option.
TODO: apply the same structure to CT-AUX, and possibly store the entire vertex configuration inside, to avoid pivoting and deleting from the back.

### CODE REVIEW

File: include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu.hpp [hunk 1]
```diff
@@ -143,7 +147,10 @@ void CtintWalker<linalg::CPU, Parameters, Real>::doStep() {
     if (configuration_.size())
       n_accepted_ += tryVertexRemoval();
     else {
-      rng_(), rng_(), rng_();  // Burn random numbers for testing consistency.
+      // Burn random numbers for testing consistency.
+      // One extra rng for the acceptance probability.
+      for (unsigned i = 0; i < n_removal_rngs_ + 1; ++i)
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu.hpp:152]: yes, it is the only reason for this class. But in the submatrix walker, it is also used to generate the appropriate number of values when building the delayed moves list.

File: include/dca/util/containers/random_access_map_iterator.hpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+// See LICENSE for terms of usage.
+// See CITATION.md for citation guidelines, if DCA++ is used for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch)
+//
+// Bidirectional iterator for the RandomAccessMap class
+
+#ifndef DCA_UTIL_TYPE_RANDOM_ACCESS_MAP_ITERATOR_HPP
+#define DCA_UTIL_TYPE_RANDOM_ACCESS_MAP_ITERATOR_HPP
+
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/util/containers/random_access_map_iterator.hpp:14]: it should. I implemented it in the library version. Should I update the PR with the library code?

File: include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu.hpp [hunk 3]
```diff
@@ -108,13 +108,17 @@ class CtintWalker<linalg::CPU, Parameters, Real> : public CtintWalkerBase<Parame
...
 };
 
 template <class Parameters, typename Real>
 CtintWalker<linalg::CPU, Parameters, Real>::CtintWalker(const Parameters& parameters_ref,
                                                         const Data& /*data*/, Rng& rng_ref, int id)
-    : BaseClass(parameters_ref, rng_ref, id), det_ratio_{1, 1} {}
-
+    : BaseClass(parameters_ref, rng_ref, id),
+      det_ratio_{1, 1},
+      n_removal_rngs_(configuration_.getDoubleUpdateProb() ? 3 : 1) {}
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu.hpp:124]: added explaining comment in dccaba53

File: include/dca/util/containers/random_access_map_iterator.hpp [hunk 4]
```diff
@@ -40,6 +40,13 @@ class RandomAccessMapIterator {
   template <class A, class B>
   using Conditional = std::conditional_t<is_const, A, B>;
 
+  // Provide STL iterator traits.
+  using iteator_category = std::bidirectional_iterator_tag;
```
Comment (PDoakORNL) [inline include/dca/util/containers/random_access_map_iterator.hpp:44]: spelling

Comment (gbalduzz) [reply include/dca/util/containers/random_access_map_iterator.hpp]: thanks for noticing it. Fixed it. Let's see if the tests pass as nothing should have changed.

**Comment by gbalduzz** [comment]
retest this, please.

**Comment by gbalduzz** [comment]
retest this please

**Comment by gbalduzz** [comment]
retest this please

**Comment by gbalduzz** [comment]
retest this, please.

**Comment by gbalduzz** [comment]
retest this, please

**Pull request merged**